[Exam] Programming exercise management

kloessst opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Programming exercise CRUD operations must be adapted for exam mode.

### Description
Client:
- [x] Import for exam programming exercises
- [x] Created new client routes for text-exercise update and creation
For creation: ':examId/exercise-groups/:groupId/programming-exercises/new'
For update: ':examId/exercise-groups/:groupId/programming-exercises/:exerciseId/edit'
For import: ':examId/exercise-groups/:groupId/programming-exercises/import/:exerciseId'
- [x] Load programming exercise module
- [x] Deactivate categories in exam mode for now, as this relies on a Course
- [x] Added "Add programming exercise" button to exercise group component  
- [x] Added exam mode to ProgrammingExerciseDetail Component. Edit should work on this page now.

Server:
- [x] Import functionality supports exam exercises now
- [x] Check that course and exerciseGroup is never set simultaneously for creation/update
- [x] Extend exercise creation/update authorization check for exam mode
- [x] Deactivate Tutor/Student notification on exercise creation/update
- [x] Add additional authorization check to GET programming exercise route. Only instructors and admins should be allowed to access an exam programming exercise
- [x] Adapted authorization check for deletion
- [x] Deactivated hints, test-cases, results-with-feedbacks and save problem statement routes for exam exercises. These routes are called by the ProgrammingExerciseEditableInstructionComponent, if the user tries to edit the exercise.

**There is a lot more functionality tied to programming exercises. For example actions available on the view page. These functions are not in the scope of this PR and will be adapted in a follow up PR.**

### Steps for Testing
_Ensure that the existing CRUD operations for regular programming exercises still work_

1. Log in as admin or instructor
2. Create a regular programming exercise
--> Exercise was created
3. Update the exercise
--> Exercise was updated
4. View the exercise
--> Can open the page
5. Delete the exercise
--> Exercise was deleted

_Ensure that CRUD operations for exam programming exercises work_

1. Log in as admin or instructor
2. Create an exam
3. Create an exercise group
4. Create a programming exercise
--> Exercise was created
5. Edit the programming exercise
--> Update was saved
6. Navigate to `/course-management/:courseId/programming-exercises/:exerciseId` using the id of an exam programming exercise
--> Should be able to see the exercise
7. Delete the exercise
--> Exercise was deleted

_Test import functionality of programming exercises_
1. Create a normal exercise -> A
2. Create a programming exercise in an exam -> B
3. Navigate to the ExerciseGroup view and click on the import new programming exercise button
--> Modal should show up with a list of programming exercises for which the user has access rights (instructor, admin)
4. Choose A for import
--> Import page should open
5. Insert the short name of A, fill out the other mandatory fields and click on import
--> Message should be displayed the the short name already exists
6. Choose an unique short name and click on the import button
--> Exercise should be imported
7. Navigate to the course exercises management page
8. Click on import new programming exercise
--> Ensure the import works by going through the steps above 

_Ensure correct authorization for programming exercises_
1. Log in as a tutor or student
2. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/programming-exercises/new`
--> You should not be able to access the route
4. Use the id of an exam exercise for the following steps
5. Navigate to `course-management/:courseId/exams/:examId/exercise-groups/:groupId/programming-exercises/:exerciseId/edit`
--> You should not be able to access the route
7. Navigate to `/course-management/:courseId/programming-exercises/:exerciseId`
--> You should not be able to see the exercise

### Screenshots

JonasPetry commented:
Codewise it looks good to me! Personally I would change the "over" in a lot of methods to "via", but I guess thats just personal preference.

kloessst commented:
@JonasPetry Noted that for later code improvements. This is not ready for review yet. I forgot to create this in draft mode. Sorry!

kloessst commented:
@pzdr7 Good findings, I will look into this

kloessst commented:
@pzdr7 @niclasschuemann 
1. I added the correct modal for deletion of programming exercises
2. For programming exercises, the route parameters are resolved inside the component and not on navigation to it. Therefor the page is displayed but not really functional. Changing this would require a rewrite of the Resolver and the handling inside the component. As this is not a issue for security, I would do this in a follow up PR, if @krusche agrees with this.
3. Fixed testing instructions

julian-christl commented on src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:
```diff
@@ -155,20 +159,28 @@ public ProgrammingExerciseResource(ProgrammingExerciseRepository programmingExer
-        User user = userService.getUserWithGroupsAndAuthorities();
-        if (!authCheckService.isInstructorInCourse(course, user) && !authCheckService.isAdmin()) {
+        // Check authorization
+        if (!authCheckService.isAtLeastInstructorInCourse(course, null)) {
```
I don't get the goal here. Why null as the user? Shouldn't fail this all the time?

anditurdiu replied:
When the user is null, the check service defaults to checking the currently logged in user

julian-christl replied:
Do you think it would make sense to overload the method to have a method 
`Boolean isAtLeastInstructorInCourse(Course course){}`
I think it's more intuitive than setting user null.

However it's not a relevant thing for release but maybe you could consider this to change in the future or reason why not.

krusche replied:
out of scope for this PR

kloessst commented:
@julian-christl I investigated the 500 error and double checked the shortname validation. Everything seems fine. I can only imagine that the Bitbucket project really exists but was already created by another Artemis instance with the same name.

julian-christl commented:
> @julian-christl I investigated the 500 error and double checked the shortname validation. Everything seems fine. I can only imagine that the Bitbucket project really exists but was already created by another Artemis instance with the same name.

Actually I wasn't able to reproduce this anymore. However were you able to find where the error is thrown in case of such an event. Because the bad error handling still exists
